Add `disable_instance` management command

stanislav-zaprudskiy opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

Add `disable_instance` management command, which allows to disable AWX
instance, optionally waiting for all its managed jobs to finish.

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

related: https://github.com/ansible/awx-operator/pull/1193

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New or Enhanced Feature

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - CLI

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 21.10.3.dev35+g3896d58f79
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```console
$ awx-manage disable_instance --wait --retry=180 --retry_sleep=10 --force-color
[32;1mInstance ["awx_1"](https://towerhost/#/instances/1/details) has been disabled[0m
1/180: Waiting 10s before the next attempt to see if the following instance' managed jobs have finished: ["sleep"](https://towerhost/#/jobs/playbook/15)
2/180: Waiting 10s before the next attempt to see if the following instance' managed jobs have finished: ["sleep"](https://towerhost/#/jobs/playbook/15)
3/180: Waiting 10s before the next attempt to see if the following instance' managed jobs have finished: ["sleep"](https://towerhost/#/jobs/playbook/15)
4/180: Waiting 10s before the next attempt to see if the following instance' managed jobs have finished: ["sleep"](https://towerhost/#/jobs/playbook/15)
5/180: Waiting 10s before the next attempt to see if the following instance' managed jobs have finished: ["sleep"](https://towerhost/#/jobs/playbook/15)
6/180: Waiting 10s before the next attempt to see if the following instance' managed jobs have finished: ["sleep"](https://towerhost/#/jobs/playbook/15)
[32;1mDone waiting for instance' managed jobs to finish![0m
[0m
```

stanislav-zaprudskiy committed:
Add disable_instance management command

Signed-off-by: Stanislav Zaprudskiy <s.zaprudskiy@sap.com>

stanislav-zaprudskiy committed:
Reformat with black

Signed-off-by: Stanislav Zaprudskiy <s.zaprudskiy@sap.com>

stanislav-zaprudskiy committed:
Support indefinitely waiting for jobs to finish

Signed-off-by: Stanislav Zaprudskiy <s.zaprudskiy@sap.com>

stanislav-zaprudskiy committed:
Do not discard jobs w/ .started=None

AlanCoding commented on awx/main/management/commands/disable_instance.py:
```diff
@@ -0,0 +1,144 @@
+        filter_group.add_argument(
+            "--hostname",
+            type=str,
+            default=socket.gethostname(),
```
The default value here is my biggest misgiving. Generally, the setting `settings.CLUSTER_HOST_ID` should be more accurate (sometimes it may get it from socket.gethostname) and will work on more types of deployments.

stanislav-zaprudskiy replied:
Updated

AlanCoding replied:
Thanks, looks good to me. I just need to understand the changes in conjunction with the awx-operator changes for a full review. This seems pretty functional now. I get the need for a "disable" command as opposed to the existing "deprovision_isntance" command. The intent is to drain jobs and disable, which is reasonable for what I know you're trying to do in the operator.

TheRealHaoLiu replied:
@rooftopcellist @gamuniz and i are running some tests with the 2 pr in conjunction right now

stanislav-zaprudskiy committed:
Use `CLUSTER_HOST_ID` as default hostname argument value

Incorporates feedback from https://github.com/ansible/awx/pull/13445/files#r1106012308

Signed-off-by: Stanislav Zaprudskiy <s.zaprudskiy@sap.com>

TheRealHaoLiu commented:
CI / api-schema failed due to PR not synced with devel 

the failure is OK bc no schema update has been introduced in this PR

TheRealHaoLiu commented:
https://github.com/ansible/awx-operator/pull/1193#issuecomment-1448791959

testing for the awx-operator side in conjunction with this PR
